### PR TITLE
gguf : enforce that tensor names are unique

### LINKED ISSUE #6836: gguf : enforce that tensor names are unique
There are models being distributed with two tensors with the same name (eg. https://github.com/ggerganov/llama.cpp/issues/6490#issuecomment-2070935873). The gguf libraries should prevent this from happening.
- `gguf_add_tensor` (ggml) and `GGUFWriter` (gguf-py) should not allow adding duplicated tensors
- `gguf_init_from_file` (ggml) and `llama_model_loader` (llama.cpp) should reject loading GGUF files with duplicated tensors, including models with multiple shards

### PR DESCRIPTION
**Pull request by ngxson**
Resolve #6836

For now, only adding duplicated tensor is prevented:
- `ggml.c` uses `gguf_find_tensor` to check for duplicated tensor. This is a linear search so may not be a suitable implementation when loading the gguf (but acceptable when writing gguf, since writing is not happen very often). What do you think about this point @slaren ?
- `GGUFWriter` from gguf-py uses a set to keep track of tensor names when writing a gguf. We can use the same logic for `GGUFReader`

### CODE REVIEW

**Comment by slaren** [comment]
A linear search is definitely not optimal, but we don't really have a good alternative at the moment, since we don't have implementations of the data structures that would be necessary to prevent this in ggml. Currently the models are small enough that it is not likely to cause a significant performance issues, and fixing the immediate issue is more important. So I think it is fine to use `gguf_find_tensor`.

File: gguf-py/gguf/gguf_reader.py [hunk 1]
```diff
@@ -234,8 +234,14 @@ def _build_tensors_fields(self, offs: int, count: int) -> tuple[int, list[Reader
 
     def _build_tensors(self, start_offs: int, fields: list[ReaderField]) -> None:
         tensors = []
+        tensor_names = [] # keep track of name to prevent duplicated tensors
```
Comment (slaren) [inline gguf-py/gguf/gguf_reader.py:237]: Is this meant to be a set?

```
AttributeError: 'list' object has no attribute 'add'
```

Comment (ngxson) [reply gguf-py/gguf/gguf_reader.py]: Yes it should be a `set()`

**Comment by ngxson** [comment]
@slaren FYI, here is the model that I used for testing: https://huggingface.co/ngxson/test_gguf_models/blob/main/stories260K_duplicated_tensors.gguf

I edited it with hex editor to have 2 times `blk.0.attn_norm.weight`

File: llama.cpp [hunk 2]
```diff
@@ -3110,9 +3110,17 @@ struct llama_model_loader {
...
 
+        std::set<std::string> tensor_names;
         for (auto & w : weights) {
             n_elements += ggml_nelements(w.tensor);
             n_bytes    += ggml_nbytes(w.tensor);
+            // make sure there is no duplicated tensor names
+            const std::string name(w.tensor->name);
+            auto found = tensor_names.find(name);
+            if (found != tensor_names.end()) {
+                LLAMA_LOG_ERROR("%s: Found duplicated tensor name %s", __func__, w.tensor->name);
```
Comment (slaren) [inline llama.cpp:3121]: It's not enough to print a message, it also needs to throw.

```suggestion
                throw std::runtime_error(format("invalid model: tensor '%s' is duplicated", w.tensor->name));
```

**Review by slaren — APPROVED** [review]
Looks good, I verified that this prevents both loading existing models with duplicated tensors, and creating models with duplicated tensors. The GPT-2 arch conversion in `convert-hf-to-gguf.py` will need to be updated to avoid duplicating `output.weight`.

**Comment by ngxson** [comment]
@slaren Thanks for the confirmation. I'm not very familiar with `convert-hf-to-gguf.py`, so unfortunately I have no clue how to do that. Without that fix, do you think this PR is still ok to merge?

**Comment by slaren** [comment]
Yes, that's a different issue, I was just pointing that's the reason why some models have duplicated tensors. Merge at will.

**Pull request merged**